### PR TITLE
Work towards green ticks in advance of Barcelona

### PR DESCRIPTION
**Pull request by fcooper8472**

### CODE REVIEW

**Comment by fcooper8472** [comment]
Just waiting for the memory testing, but assuming that goes green, this PR is ready to merge. This should give us a nice green tick on the landing page.

**Comment by fcooper8472** [comment]
Summarising the key changes in this PR:

- Overhaul the list of clang tidy static analysis checks being run
- Increase the amount of testing we're doing, to match what Buildbot was doing
- Turn on C++17 and bump the cmake minimum required to match Ubuntu 20.04 (#107)
- Do the conditional compilation with `if constexpr` and `[[maybe_unused]]` (#107)
- Tidy up our testing related to division by zero

File: heart/test/postprocessing/TestCellProperties.hpp [hunk 1]
```diff
@@ -97,7 +97,7 @@ class TestCellProperties : public CxxTest::TestSuite
 
             //Should throw exceptions because the cached vector of onset times (mOnsets) is empty
             TS_ASSERT_THROWS_THIS(cell_properties.GetLastActionPotentialDuration(90), "AP did not occur, never exceeded threshold voltage.");
-            TS_ASSERT_THROWS_THIS(cell_properties.GetAllActionPotentialDurations(90)[0], "AP did not occur, never exceeded threshold voltage.");
```
Comment (mirams) [inline heart/test/postprocessing/TestCellProperties.hpp:100]: why's this changed?

**Review by mirams — APPROVED** [review]

**Comment by fcooper8472** [comment]
Note that work resolving the problems with the Intel compiler has been merged into this PR from #112

**Comment by fcooper8472** [comment]
@mirams @kwabenantim @BJackal @AlexFletcher @MILeach @jmpf I've added in all the necessary fixes for Intel, so this makes the full build clean and green.

@mirams was happy with the PR pre-Intel changes, but if someone else fancied taking a look that would be helpful.

If you're reviewing the changes, you can click the cog at the top and there's a tickbox to "Hide whitespace", which makes the `if constexpr` stuff mostly disappear: you only see the actual line changes, not all the indentation.

**Review by jmpf — APPROVED** [review]
I reviewed on Monday and at Tuesday's meeting (all but the Intel NaN exceptions).  Recent changes are also good.

**Pull request merged**